### DISCUSSION TITLE
How to scale properly world to canvas

### DISCUSSION THREAD

**Discussion by mwpowellhtx** [mono/SkiaSharp#2998] [Q&A]
Hello,

I'm having what I think is an odd precision issue converting from world coords to canvas bitmap.

Assumptions. The canvas bitmap has a resolution, presumably some pixels `width` and `height`, `int`. Although from a canvas perspective, these are, I think, in terms of `float`, correct? That's first.

Second. _world coordinates_ are intentionally tracked in `decimal`, because precision is critical. Measurements are to the _centimeter whole number_, including _fractional decimal places_ for smaller than that. So cannot have `float`, or especially `double`, floating point rounding precision errors going on.

Third. _World true zero_ is not canvas `{0,0}` presumably top left, usually. There is a simple offset `{dX,dY}` getting to a zero world coordinates that is aligned with zero canvas. Neither here nor there, I think, for purposes of discussion, just an interesting translational mad fact.

![image](https://github.com/user-attachments/assets/cd30ae9b-6921-461f-9a7b-0e9211e84b50)

Anyway, at the moment, I have this translation going on from `Location` (3D, including `Z`, but we do not care about that dimension, `decimal`) to `SKPoint` (`float`). Which I feel is scaling in correctly, which I suspect may be due to some precision issues between `decimal` through `int` to `float` or some parts in between.

![image](https://github.com/user-attachments/assets/34538bcf-31c0-46a7-8542-ff987c41460e)

The basic algorithm is to assess the _ratio of world coordinate in world dimensions_, and apply that into the _canvas bitmap dimension_. The fun part is, the algorithms and I think scaling to SKIA points is at least consistent. But the location I can assure you, precision wise, is off.

```csharp
public static SKPoint ToSkiaPoint(this SKBitmap bitmap, Location position)
{
    var worldToCanvasOffset = WorldToCanvasZeroOffsetCentimeters;

    var canvasPos = position + worldToCanvasOffset;

    // Take only two leaving out the Z coord.
    var coords = canvasPos.Coordinates.Take(2).Select(_ => _.Value).ToArray();

    // We need the precision in terms of defimal or even float at this point.
    (decimal bwidth, decimal bheight) = (bitmap.Width, bitmap.Height);
    (decimal wwidth, decimal wheight) = WorldSizeCentimeters;

    // Here 'last' is not Z but rather the Y coord.
    (decimal x, decimal y) = (coords.First() / wwidth * bwidth, coords.Last() / wheight * bheight);

    return new((float)x, (float)y);
}
```

So the algo goes something like:
1. Translate `position` to _canvas friendly zero_ `{0,0}`.
1. Take the first two `{X,Y}` _coordinates_ from the _position_.
1. Identify the _bitmap_ and _world width_ and _height_ components.
1. Apply the appoprirate _ratio_ to each of the coordinates. Ratio should be generally, "bitmap dims per world dims".
1. This should be our `SKPoint`.

I think maybe I see the issue, possibly. Perhaps I need for all the components to land in `float`, for smoothest possibly scaling into canvas `SKPoint`.

My next steps are to decompose the calculation a bit and better assess things. But if there are any tips how to perform this calculation more precisely, I am all ears.

**Comment by mwpowellhtx** [comment]
Although I will say after reading couple of background blogs, maybe there is indeed an easier way to train the canvas as to the world coordinates, and let it handle those translations and scalings for us; which would be beneficial, because then, potentially, we could simply make the direct jump from our world coordinates straight through the canvas. POTENTIALLY...

**Comment by mwpowellhtx** [comment]
I should probably also mention, unlike the academic scaling transformative examples, in which simple shapes are drawn on a blank canvas, I am loading a reference file from drive, on which I want to apply some scaled (to that bitmap resolution) area, translated then scaled from world coordinates. That's a pretty non-trivial difference, IMO, and something that should be taken into consideration. The reference file is a _"hard fact"_ non-negotiable in that equation. Without which the properly translated and scaled layers above which are meaningless.

In very high level 10K meter perspective; I am adding layers to a map; in other words, it is a cartography problem. And the map, although more or less fictional in nature, is a hard fact about that problem.

Hope that helps, if any one is thinking of making any suggestions toward solving the present dilemma.

Cheers, thanks... 🍻

**Comment by mwpowellhtx** [comment]
I adjusted the algo slightly, but still appears to be a `decimal` to `float` precision issue going on scaling into the `int` based _canvas bitmap_.

```csharp
// SKPointM is decimal based, we do the world-to-canvas translation shifting canvas friendly {0,0} zero.
var pointM = position.ToPointM();
var (px, py) = ((float)pointM.X, (float)pointM.Y);

// We need the precision in terms of defimal or even float at this point.
var (bwidth, bheight) = (bitmap.Width, bitmap.Height);
//(decimal bwidth, decimal bheight) = (bitmap.Width, bitmap.Height);
var (wwidth, wheight) = ((float)WorldSizeCentimeters.Width, (float)WorldSizeCentimeters.Height);
//(decimal wwidth, decimal wheight) = WorldSizeCentimeters;

// Here 'last' is not Z but rather the Y coord.
var (x, y) = (px * (bwidth / wwidth), py * (bheight / wheight));
// Best if thought of as "times 'bitmap per world'" units, cancels the world units, should be landing in bitmap units.

return new(x, y);
```

The values in question in the debugger.

```
(float px, float wwidth, int bwidth, float bitmapPerWorldX, float x) = (451915.531, 1500000, 16384, 0.0109226666, 4936.12256);
(float py, float wheight, int bheight, float bitmapPerWorldY, float y) = (993356.9, 1500000, 16384, 0.0109226666, 10850.1064);
```

Our baseline cartography bitmap is `16384x` resolution. As stated earlier, that is a _"hard fact"_, and whatever else we do, that scale cannot be mutated.

The world coordinate system is `1500000x` `decimal`, that, too is a fact, but we want to precisely fit the location into the overlay.

The strategy is to apply the _"bitmap per world"_ unit conversion, which we think should be canceling out the world, and arriving in terms of (hopefully precise) bitmap canvas positions.

I've tried to calculate with `decimal` through and through, convert to `float` at the very end, no difference. Or `float` through and through. No difference, still there is some sort of scaling issue going on.

I'm almost wondering is there an aspect ratio thing going on, but that does not seem quite as likely.

**Comment by mwpowellhtx** [comment]
So far what I have tried...

1. Do calculations in terms of `decimal` until the last possible `(float)coord` moment.
1. Do calculations in terms of `float`.
1. Do calculations in terms of `double` until the last possible `(float)coord` moment.

None of which seem to be making the least bit of difference. The core issue, is that I want to cram a large world coordinate space very precisely into a smaller bitmap resolution.

The next thing I will try for is to perhaps start with a larger resolution source bitmap, which I have several, one in particular extremely large, but really, for performance issues, trying to avoid that if I can help it.

**Comment by mwpowellhtx** [comment]
Next tried loading from higher resolution baseline image, `32Kx`, does not improve things.

Will investigate the efficacy scaling the baseline image out to `1:1` _world:canvas_ if at all possible. And maintain a reference to that baseline bitmap. Might also help improve overall performance not having to load the file every time. Depending what SKIA does expanding the bitmap into a larger bitmap canvas resolution.

**Comment by ojb500** [comment]
I can't see any big problem with what you are doing. What is going wrong exactly?

I wouldn't worry too much about `float` precision, if it is a real concern then transform your coordinates so they all lie near 0 where float has maximum dynamic range, but if you're drawing to a bitmap there is a bound on how precise you can possibly be (the equivalent of 1 pixel in world space)

**Reply by mwpowellhtx** [reply]
There's only so much I can do re: transformations. I do the translation from world zero to canvas zero at the outset, i.e. `ToPointM`.

In terms of 'world' or more specific 'view ports', yeah, that is another idea. Bit more complex to think about, but certainly another path to consider.

Trying to resize the image itself to anything approaching `1.5M` resolution, non-starter, Skia balks at that. Tried even `{w*w, h*h}`, from `16K` res or so, also non-starter Skia exception.

So starting to wonder the view port approach FTW. Then could potentially scale that out to a closer 1:1 world:canvas resolution.

**Reply by mwpowellhtx** [reply]
To clarify, as the first image suggests, the circle and points are drawn relative to each other in the correct geometries. However, the overall map position is off, as indicated by the arrow, should be shifted somewhat to the left. Which I think is the resolution scaling issue.

**Reply by mwpowellhtx** [reply]
The very tricky part is that the source image is rotated 180 degrees from what it needs to be in the final product. I suppose could also rotate the view port 180 degrees and do the appropriate view translations accounting for view top left and just do it that way. Then also rotate the view when ready to output, present, etc.

**Reply by mwpowellhtx** [reply]
Turning the problem over in my mind's eye, still I think there must be some function of the view scaling, world:canvas that  can be calculated and translated for. Because any way you slice it, jumping through that first initial bitmap, the view scale there in, there will al ways be that inaccuracy. Does not matter what I am scaling for; rectangle edges i.e. for that view port, circles, points, whatever.

**Reply by mwpowellhtx** [reply]
I will also add, I also apply layers for grid lines and sector keypad labeling. These are quite precise, but I am also not leaning into the world coordinate system to accomplish that, but rather the bitmap dimensions themselves, and given some forward knowledge on the expected rows and columns in the grid, overlayed keypad, in the appropriate orientations and so forth.

**Comment by mwpowellhtx** [comment]
Following up, there is [_this reference_](https://skia.org/docs/user/coordinates) in the SKIA docs, but I do not know what the `CSharp` analog might be, I do not see a `SKRect.MakeXYWH`, for instance, which to me might see the best ideal letting SKIA do the heavy lifting, if at all possible.

Then there is also this [_GPT_](https://math-gpt.org/3379928b-c8c1-4712-8a8f-43f6ed14205a) dialog I have been adapting, somewhat, trying to drill into potential problem areas of concern. And while I am loathe to _"let"_ any AI direct my coding, if it can inform me the strongest, most robus algorithm, efficacy, pros cons, then why not.

But still, I am not hundred percent sold I can repeatably determine the post-scaling translation consistently. The chief opportunity is, that whenever I want to jump through the world lens to canvas for any reason, there is always a progressively worser post-scaling tranlation issue.

One case in view, pardoning the pun, from `1.5M` _world dimensional length_, that is a hard constraint, to `8192x` _bitmap image dimention_, effectively about `1/183` scaling `world:canvas`. Of coruse that ratio can vary with more or less dimensional length; `~64Kx` in the most extreme, not taking into consideration load on CPU, RAM, etc.

So, bottom line, I am after the best possible algorithm. Maybe the third exchange with my GPT session is the most promising. The poly-regression approach might be a bit overkill, I'm not sure yet.

Then there is the matter of how I may facilitate such a calculation; which as I understand it [`System.Numerics`](https://learn.microsoft.com/en-us/dotnet/standard/numerics) matrices might be the way to go there, however, can be tedious to work with matrices at any level.

Anyway 🍔 🌭 🍕 for 🧠  ...

**Comment by mwpowellhtx** [comment]
Maybe I am doing too much work there as well thinking about it. If I can simple lean into the [`SKCanvas.Scale`](https://learn.microsoft.com/en-us/dotnet/api/skiasharp.skcanvas.scale) method. Are there examples how to use this? That would be grand if that handled the nasty details for us, especially when points are getting less precise because the world coordinate system is larger than the bitmap image resolution. And we could simply concern ourselves with the drawing parts, either in world (ideally) or canvas coordinates. Worse case there is also SKIA matrix support, if necessary, but might like to avoid that scenario.

**Reply by mwpowellhtx** [reply]
Actually I wonder if that is something else entirely... Like as in it is drawing a scale on the map. Although I do not see any `Paint` choices there; but which would be great nonetheless; albeit not exactly what I am talking about here.

**Reply by mwpowellhtx** [reply]
Okay so not drawing a map scale, but from this [_SKIA fiddle example_](https://fiddle.skia.org/c/@Point_scale), it is rather academic, some things need to be properly understood, like is that scale of `world:bitmap`... so if we know we have `1.5M` centimeters in a dimension, and a _bitmap resolution_ of `8192x`, then we can say _px/centimeter_ or `1/183`, or this ugly floating point value (_Windows Calculator_) `0.00546448087431693989071038251366` albeit in single precision `float`, so more like akin to `0.00546f` perhaps... which probably accounts for some of the drift I expect to see the further out we go. Even if we set _bitmap-center_ as the _scaling center point_, maybe that mitigates somewhat. Then do we also need to account for `X:Y` aspect ratios. Overall the docs are stone cold silent on the `Scale` method itself.

**Comment by mwpowellhtx** [comment]
Okay so this thing adapting `Scale` in, I'm not sure where it is drawing, if at all. I see no evidence that it is. Or that Scale is otherwise being respected throughout, at least not based on the academic fiddle example (above). That's first off.

```csharp
public static void VerifyCanvasScaleDryRun(string imagePath, Guid uuid = default)
{
    string fullPath;

    if (uuid == default)
    {
        uuid = Guid.NewGuid();
    }

    fullPath = $"Maps\\canvas_scale_dryRun_{uuid = Guid.NewGuid()}.jpg";

    using var image = SKImage.FromEncodedData(imagePath);
    using var bitmap = SKBitmap.FromImage(image);
    using var canvas = new SKCanvas(bitmap);

    var bitmapWidthM = (decimal)bitmap.Width;
    var bitmapWidthF = (float)bitmap.Width;
    var bitmapWidthD = (double)bitmap.Width;
    var scaleM = bitmapWidthM / WorldSizeCentimeters.Width; // 8192 / 150000 in our use case...
    // TODO: this is a problem... Windows Calculator reports: 0.00546448087431693989071038251366
    var scaleF = bitmapWidthF / (float)WorldSizeCentimeters.Width;
    var scaleD = bitmapWidthD / (double)WorldSizeCentimeters.Width;
    // TODO: but single/double are reporting something more like this: 0.0054613333333333337
    canvas.Scale(scaleF);

    // TODO: pg record: {"Uuid":"b8ec5136-c7e8-4e7b-bfc4-ab26d394c8c2","X":-438084.469,"Y":103356.875,"Z":38187.777}
    const float radius = 10000f;
    using var paint = new SKPaint { Color = SKColors.Red.WithAlpha((byte)(byte.MaxValue * 0.5f)) };
    SKPoint point = new(-438084.469f, 103356.875f);
    canvas.DrawCircle(point, radius, paint);
    canvas.Flush();

    const float inflation = radius * 10;

    SKRect rect = new((point.X - radius) * scaleF, (point.Y - radius) * scaleF, (point.X + radius) * scaleF, (point.Y + radius) * scaleF);
    rect.Inflate(inflation * scaleF, inflation * scaleF);

    var bitmapRectI = bitmap.GetBoundsRectI();
    var viewBitmap = new SKBitmap((int)rect.Size.Width, (int)rect.Size.Height);

    using var viewCanvas = new SKCanvas(viewBitmap);
    viewCanvas.DrawBitmap(bitmap, bitmapRectI, viewBitmap.GetBoundsRectI());

    // TODO: canvas scale makes it look very grainy pixelated... not gonna work for us...
    // TODO: image quality is EVERYTHING
    using var data = bitmap.Encode(SKEncodedImageFormat.Jpeg, 100);
    using var fs = new FileStream(fullPath, FileMode.Create, FileAccess.Write, FileShare.Read);
    data.SaveTo(fs);
}
```

Secondly, there are definitely _floating point rounding_ issues going on. I'm not sure quite what we can do about that, which is going to make dealing with some resolutions very challenging when `decimal` centimeters accuracy precision is literally EVERYTHING to the cartography question.

Just on the numbers alone, I'm not sure how they get `0.0054613333333333337` from `0.00546448087431693989071038251366` from about the sixth decimal point forward. Maybe there is a precision numerics solution to that, I do not know yet. Or that we might even think about kicking it out to a C++ library where we do have precision, and loading that result. Even taking the raw `float` numbers, not converted from `decimal`, the result is not better, `testF = 0.00546133332 /* 8192f / 1500000f */`.

Drawn with scaling involved, still the same sort of floating point rounding issues evidenced. The only up side is that we can more than less specify geometry in terms that are literally world scale. The following example, not rotated, but still very inaccurate, manifestation of the rounding issues.

![image](https://github.com/user-attachments/assets/bf46b31f-40ca-488b-a58e-db70f8e5c436)

There are also issues capturing view ports when scaled, result is VERY pixelated. Maybe need to reset scale to `1f` approaching that bit, I do not know. Somewhat separate issue, the rounding issue is the core problem right now.

Will experiment with higher resolution versions of the source map to see if and how it may improve. I would expect some, but kind of at a loss how to achieve accuracy at this point.

**Reply by mwpowellhtx** [reply]
Even the decimal version of the same scaling appears to be trash beyond about the 6th place in the debugger. Do not know if that is just the debugger or what the issue is, `0.0054613333333333333333333333M`.

**Reply by mwpowellhtx** [reply]
And for giggles, rendered to string `"M = 0.0054613333333333333333333333, S = 0.0054613333, D = 0.005461333333333334"`.

```csharp
var scaleString = $"M = {scaleM}, S = {scaleF}, D = {scaleD}";
```

**Reply by ojb500** [reply]
I think you probably don't want to scale the canvas itself, this is just applying a transformation to any future drawing commands you would make. 

There is an important distinction that should be drawn between the canvas and the underlying drawing surface. An `SKCanvas` does not necessarily draw into an `SKBitmap` (although this is of course very common). There are `SKCanvas`es that [generate SVG drawing instructions](https://learn.microsoft.com/en-us/dotnet/api/skiasharp.sksvgcanvas?view=skiasharp-2.88) or [record instructions to be played back later](https://learn.microsoft.com/en-us/dotnet/api/skiasharp.skpicturerecorder.beginrecording). The SKCanvas just provides the API to let you draw on the surface, whatever that may be.

Try to think in terms of 'world space' and 'screen space' (or 'pixel space'). It seems like you deal with a fundamentally square 'world' and have a square 'screen' that you output to as well, so any conversion between the two spaces is just some combination of translation and scaling.

You can try to write two functions 
`Location ScreenToWorld(SKPoint pt)`
`SKPoint WorldToScreen(Location loc)`

Test these functions. Are they inverses of each other (to adequate precision)? Do you get a sensible result for e.g. `ScreenToWorld(new SKPoint(0,0))` (presumably this should be the 'top left'/northwestmost point of your world)? How about the bottom right corner, is this the southeastmost point of your world? If you call `WorldToScreen` with your zero point, is this in the right place?

Another word about precision. If you have a 150-meter-wide world that has to fit into a bitmap of 8192 pixels wide, then 1 pixel equates to 150m/8192 = 18.3cm. You can't draw any more accurately than that because a pixel is either lit up or not, so 18.3cm is the maximum accuracy you can achieve at this resolution regardless of how precise your input coordinates are. (I work with systems that have to draw the whole world at scales from 1:30m down to 1:a few hundred, and there is no problem working with `SKPoint`s represented as pairs of `float`s provided you're a little careful about when this conversion occurs - we want to lose any precision at the last possible moment when we are already working with relatively small numbers that are near screen space)

**Reply by mwpowellhtx** [reply]
Yes would like to avoid scaling the canvas. It can be done, and has the benefits above. But the after effects, pixelation, are disastrous.

Ostensibly, yes, that is my objective, to functionally calculate `WorldToCanvas`, as you suggested.

Have tried now several different approaches involving the scaling factors, contending with dotnet floating point precision, BigDecimal 3P packages, and so forth. And anyway I slice it, the crux of the issue is the scaling factor alone is insufficient. I am missing a translational factor post-scaling.

There are solutions for that, apparently, but involves higher order, though not impossible to comprehend, linear algebra.

To the points about translation and scaling; there are two translations going on. First is world to canvas-friendly. I handle that front side of the canvas geometry. So can assume world zero aligns with canvas. That aside...

It seems like to me there is a scale-translation, missing the translation component that should apply however I slice it. The scaling is always off, regardless of the precision used. Using a package called `ExtendedNumerics.BigDecimal` to facilitate the scaling calculations.

The results are always consistently off.

![image](https://github.com/user-attachments/assets/48898191-a888-4264-9f62-430e0ec8428c)

But thinking I need the matrix approach given known scaling factors, unknown translational components.

**Reply by mwpowellhtx** [reply]
[_The GPT result_](https://math-gpt.org/3379928b-c8c1-4712-8a8f-43f6ed14205a), second one involving solving for translational components.

**Reply by ojb500** [reply]
Yes, scaling the canvas will lead to undesirable artifacts when you draw your backdrop. As I mentioned though, arbitrary precision math really shouldn't be necessary for this, so I think you must be missing something more fundamental.

Here's a simplified picture of what I think you are trying to do here (I've used smallish, round numbers for sanity's sake)

![image](https://github.com/user-attachments/assets/c13ca6e7-781c-43f1-9bfc-7884cd7975a9)

We have a world that goes from -75 to +75 metres with its zero point at the centre. A total of 150m wide.
Our canvas is 100px wide and has its origin in the top left according to skia conventions.
We would like to map blue to blue, red to red, black to black.

We map a point from world to screen by
* Shifting the origin to the top left corner by subtracting -75 from both coordinates.
* Multiplying the coordinates by the ratio 100/150 (the size in pixels divided by the number of world units)

Mapping a point from screen to world goes in reverse:
* Multiplying the coordinates by the ratio 150/100 (the number of world units divided by the size in pixels)
* Then we move the origin back to the origin of the world space by adding -75 to both coordinates.

Try to go through these steps and visualise how the rectangle changes as we do each operation. Convince yourself that this works :) There are no unknown components or higher order algebra in this. If your spaces are not squares this still works but you will need to use different scaling factors along each dimension.

When we do scaling and translation operations like this to map one rectangle to another, we are doing an 'affine transformation' (to use the geometric term) which can also be represented as a transformation matrix if you're more of a linear algebra guy. But really it's just addition and multiplication, nothing too high level required.

**Reply by mwpowellhtx** [reply]
As far as whether the actual scaling output is accurate; no it is not. It is off, evidenced by the above image. Nearer to the middle of the source image, the error is less, oddly enough. But further out, seems very much less accurate.

Just had another interesting observational breakthrough, which will make coordinates surveys interesting, BUT...

When I feed the algorithm with VERY precise coordinates, the scaling-only approach seems good.

```csharp
SKPointM pointM1 = new(-4380.84469m, 1033.56875m); // INACURATE, meters
SKPointM pointM2 = new(-4309.5458984375m, 1007.8369140625m); // PRECISE, determined by Windows Calculator
SKPointM pointM3 = new(-8398.291015625m, 3318.6279296875m); // PRECISE, ditto Calculator
```

![image](https://github.com/user-attachments/assets/6bf2033b-8f03-4719-8ae6-e6521025962b)

And the other test coordinates, also spot on. But only when I started from highly precise coordinates. In this case, placed exactly where I wanted it to, on the building near the radar array.

![image](https://github.com/user-attachments/assets/c3d736ca-be82-474a-a487-bdeb970fbebe)

But short of precision coordinates surveys, the crux of it is an imprecise scaling factor + post translation.

**Reply by ojb500** [reply]
> As far as whether the actual scaling output is accurate; no it is not. It is off, evidenced by the above image. Nearer to the middle of the source image, the error is less, oddly enough. But further out, seems very much less accurate.

Is the middle of the source image where your world zero is? Then most likely, your scaling factor is not correct.

**Reply by mwpowellhtx** [reply]
In reference to your illustration, the issue is not the math itself. The issue is the precision of the world:canvas scaling factor. But as noted above, when I start from very precise coordinates, like further than 3-places or 5-places, depending whether centimeters or meters, then the result is far closer, if precise, as expected.

**Reply by mwpowellhtx** [reply]
> > As far as whether the actual scaling output is accurate; no it is not. It is off, evidenced by the above image. Nearer to the middle of the source image, the error is less, oddly enough. But further out, seems very much less accurate.
> 
> Is the middle of the source image where your world zero is? Then most likely, your scaling factor is not correct.

I translate world to a zero that is in the same place as the canvas.

Actually, the world grid is `1,500,000`, and zero is more like `890,000`. Not the true center of the bitmap image. And the coordinates are rotated 180 degrees, if that makes sense.

But those transformations are neither here nor there. For math sanity, we start from the rotated bitmap image, then rotate when we are ready to present the final result.

**Reply by ojb500** [reply]
Can you share some of your `SKPointM` code? `decimal` may not be a good choice for doing this kind of math.

**Reply by mwpowellhtx** [reply]
> Can you share some of your `SKPointM` code? `decimal` may not be a good choice for doing this kind of math.

Basically a `float`-`decimal` conversion from the `SkiaSharp` code. `decimal` is a must because, technically, notwithstanding dotnet floating point rounding shenanigans, whether in `centimeters` or `meters`, precision is a must. Cannot have any rounding issues that would influence the results.

So... whereas `decimal` is a great fit for financial applications, for instance, because of that. It is also a must have for this one, for the aforementioned reaons. We basically do a `float` conversion, or `BigDecimal`, when doing the calculations, as the case may be.

But were also experiencing floating point rounding issues beyond 5-6 places. Troubling to say the least.

**Reply by ojb500** [reply]
`decimal` is still a floating point type and you can still accumulate errors if you're not careful - it's just bigger than the others and uses 10 as the base rather than 2. It's very good at representing *decimals* but might not be as good as `double` for some cases (see what happens when you do `Console.WriteLine(1M / 3M * 3M);` vs `Console.WriteLine(1.0 / 3.0 * 3.0);`)

When you calculate your scaling factor this should be done with fairly wide precision (probably `double` is more than adequate as the number will be small) and the subsequent multiplication of the coordinates should also be done in wide precision. Only then (when you're dealing with pixel positions, small-ish numbers between 0-8192) should you convert your coordinates to `float`s and make an SKPoint. A `float` has no problem representing any integer up to about 16 million or so.

**Reply by ojb500** [reply]
See [this fiddle](https://sharplab.io/#v2:C4LgTgrgdgNAJiA1AHwAICYAMBYAUBgRjzwDcBDMAAgDMAbAe0oF5KAOAgTnQDpNqB6AgFZMo0b2oBuPKk4AKACQAiAGIMywSgGcAxmVoBLKAHMaZHcHpgQlAN516AXyUBKabnJU4EAEbM2nDyYgiJimLzushyKSgAi9L60AKbaeoYmZhZWNrbePs5upBSUcEk6/uxcALIhYZhVkfLKsWUGALb6qfpGptTmltZ2pToF7jJNSgDqVrRwlEkAHsBJUJo+ZFpJc/RQNOqaut0ZfVmDtpXo/A6j49HK02Cz80sraxtblDslCT7JXem9frZOwXfh5G74CYPJ6LZarSjrTbbXbDdqdQ4AzIDHKg4YQoA==) - `decimal` seems to give the most significant digits in the scaling factor, but leads to a surprising result.

**Reply by mwpowellhtx** [reply]
An interesting observation but for different reasons, you are verifying what we were seeing in the `CSharp` discords earlier.

```
Float scaling factor: 0.0054613333
Double scaling factor: 0.005461333333333334
Decimal scaling factor: 0.0054613333333333333333333333
```

All things considered, maybe the scaling factor _"is fine"_.

But actually, I think the issue may be rooted in the source coordinates themselves, [_as noted above_](#discussioncomment-10675601).

Unconfirmed, but the thing reporting the coordinates, I'm not sure may be truncating or rounding itself, and maintaining more precise numbers internally. Although I cannot be certain. When I do the math starting from an image editor, and ratios on screen, then the result is what is expected.
